controller/fifo_cache.go

```diff
@@ -0,0 +1,62 @@
+package controller
+
+import "sync"
+
+// fifoCache is a very simple first in first out in-memory cache storing string
+// pairs.
+type fifoCache struct {
+	data      []string
+	dataMutex *sync.Mutex
+}
+
+func newFifoCache(size int) *fifoCache {
+	if size < 1 {
+		// Yes this panics. But if size is less than 0 it will panic
+		// anyway in slice creation.
+		panic("size must be bigger than 0")
+	}
+
+	f := &fifoCache{
+		data:      make([]string, 0, size),
+		dataMutex: &sync.Mutex{},
+	}
+
+	return f
+}
+
+// Put inserts and item into the cache.
+func (f *fifoCache) Put(s string) {
+	f.dataMutex.Lock()
+	defer f.dataMutex.Unlock()
+
+	if f.containsThreadUnsafe(s) {
+		return
+	}
+
+	if len(f.data) != cap(f.data) {
+		f.data = append(f.data, s)
+		return
+	}
+
+	copy(f.data, f.data[1:])
```

Comment:
Is this memory efficient or does that leak somehow? I remember there are some slice tricks that appear to be better in certain situations.

Reply:
Cache implementation changed.